### ISSUE TITLE
Apple Silicon support

### ISSUE THREAD

**Issue by iamahuman** [AndreRH/hangover#97]

**Comment by stefand** [comment]
Neither of us has the necessary hardware yet, and 16kb pages will be a major problem.

**Comment by darkbasic** [comment]
> Neither of us has the necessary hardware yet, and 16kb pages will be a major problem.

You don't have to, using 4K is as simple as:

```
CONFIG_ARCH_MMAP_RND_BITS_MAX=33
CONFIG_ARCH_MMAP_RND_BITS_MIN=18
CONFIG_ARCH_MMAP_RND_COMPAT_BITS_MAX=17
CONFIG_ARCH_MMAP_RND_COMPAT_BITS_MIN=11
CONFIG_PPC_4K_PAGES=y
# CONFIG_PPC_64K_PAGES is not set
CONFIG_PPC_PAGE_SHIFT=12
CONFIG_FORCE_MAX_ZONEORDER=13
CONFIG_ARCH_MMAP_RND_BITS=18
# CONFIG_VMXNET3 is not set
```

The previous snipped is for ppc64le, but you can do the same on M1 because 4K is supported natively as well. In fact Rosetta is a 4K application: in MacOS 16K and 4K can coexist.

**Comment by stefand** [comment]
Yeah but following the Asahi Linux people's statements, while the CPU supports 4k pages, the iommu does not. Linux will have to gain support for per-process page sizes, similar to macos, or will have to deal with a CPU/IOMMU page size mismatch. That's some time away.

**Comment by darkbasic** [comment]
AFAIK in 2021 they were on their way to fix that issue:

```
IOMMU 4K patches (in review): The M1 is peculiar in that, although it supports OSes that use either 16K or 4K pages, it really is designed for 16K systems. Its DART IOMMU hardware only supports 16K pages. These chips have 4K support chiefly to make Rosetta work on macOS, but macOS itself always runs with 16K pages – only Rosetta apps end up in 4K mode. Linux can’t really mix page sizes like that and likely never will be able to, so we’re left with a conundrum: running a 16K kernel makes compatibility with older userspace difficult (chiefly Android and x86 emulation), plus distros don’t usually ship 16K kernels; while running a 4K kernel runs into a major mismatch with the DART. This initially seemed like a problem too intractable to solve, but Sven took on the challenge and now has a patch series that makes Linux’s IOMMU support layer play nicely with hardware that has an IOMMU page size larger than the kernel page size! It’s not perfect, as it can’t support a select few corner case drivers (that do things that are fundamentally impossible to support in this situation), but it works well and will support everything we need to make 4K kernels viable.
```

Is there any news on that?

**Comment by stefand** [comment]
No real progress on the page size thing: https://arstechnica.com/gadgets/2023/02/linux-is-not-exactly-ready-to-run-on-apple-silicon-but-give-it-time/

**Comment by darkbasic** [comment]
Apparently things are finally moving: 

> FEX doesn’t work on standard Asahi Linux kernel builds since we use 16K pages, but 4K page support is not actually that difficult to add… so starting this week, I’m going to be adding 4K support to the Asahi GPU driver and fixing whatever issues I run into along the way, and then we’re going to try running Steam and Proton on it!

https://asahilinux.org/2023/03/road-to-vulkan/

**Comment by ToCodeABluejay** [comment]
Perhaps if you try a custom kernel build? Will have to give it a try...🤔https://www.reddit.com/r/AsahiLinux/comments/tq575y/possibility_to_install_a_4k_kernel_on_arch/

**Comment by ToCodeABluejay** [comment]
https://github.com/AsahiLinux/linux/issues/47?fbclid=IwAR2EEOuENLoK9P3QT9gomHIPm31yH1pZrdVXbPb9aH5V5xR8WAQzCQvsvZ4#:~:text=That%20being%20said%2C%20it%20does%20work%20but%20the%20implementation%20is%20a%20bit%20of%20a%20hack%20and%20I%E2%80%99m%20waiting%20for%20another%20series%20by%20Robin%20Murphy%20which%20cleans%20up%20the%20iommu%20API%20before%20I%20continue%20working%20on%20it.

**Comment by ToCodeABluejay** [comment]
After successfully installing llvm-mingw on Asahi, and building and installing a 4kb kernel, I tried building hangover only to get the following build failure...

       ld.lld: error: undefined symbol: ___wine_import_ucrtbase_name
        >>> referenced by /tmp/libucrtbase-00000000-80cda7.o:(___wine_import_ucrtbase_desc)

        ld.lld: error: undefined symbol: _WinMain@16
        >>> referenced by ../crt/crtexewin.c:67
        >>>               libmingw32.a(lib32_libmingw32_a-crtexewin.o):(_main)
        clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
        winebuild: /usr/bin/i686-w64-mingw32-as failed with status 1
        make: *** [Makefile:157954: dlls/ucrtbase/i386-windows/libucrtbase.a] Error 1
        make: *** Attente des tâches non terminées....

Building with the following release of llvm-mingw-ucrt on Asahi Linux using edge kernel 6.3.0-5
https://github.com/mstorsjo/llvm-mingw/releases/tag/20230603

**Comment by Matthew-Bradley** [comment]
> Apparently things are finally moving:
> 
> > FEX doesn’t work on standard Asahi Linux kernel builds since we use 16K pages, but 4K page support is not actually that difficult to add… so starting this week, I’m going to be adding 4K support to the Asahi GPU driver and fixing whatever issues I run into along the way, and then we’re going to try running Steam and Proton on it!
> 
> https://asahilinux.org/2023/03/road-to-vulkan/

Sadly, this doesn't include any work to fix the IOMMU problem.

I think they looked at reworking core linux code to address the IOMMU, but have not worked on it. Part of that is because the work is extremely invasive and it would be a ton of political work to get linux to accept the patches. But, I think they might also be trying to encourage applications (like chromium, for example) to fix their 4K page assumptions and get better at running on other page sizes.

Sooner or later I guess some kind of solution will need to be found for emulating crusty old 4K x86 code on future computers with larger page sizes. That's probably out of scope for hangover though.

**Comment by stefand** [comment]
There are hypothetical solutions to the 16kb page size problem.

Windows might use 4kb page sizes, but the API mostly exposes 64k page sizes. VirtualAlloc et al work on 64k boundaries. The reason for this is the historic support for Alpha CPUs on early Windows NT.

The place where you get 4k granularity is mapping PE files and VirtualProtect on existing allocations. Wine's own PE files can be built with 16kb section alignment. I suspect (but don't have any sort of proof) that most applications will be happy by using the union of access permissions on 4 4kb pages that get mapped into one 16kb page (i.e., if 8 kb of read-write .data and 8kb of read-execute .code get mapped into one place, make it a 16kb .rwx page). It might be possible to relocate individual sections if the file contains relocation symbols (although the increase in PIC code means that more and more DLLs don't have reloc symbols because they normally don't need them).

Another option is to to give a merged page minimum permissions (read-only in the above example), catch exceptions and manually handle writes and execute attemps. When an emulator is involved the emulator might be able to help here. This approach has its own problems (e.g. we might have to single-step through code that shares a page with non-executable data), performance may suffer, etc.

And finally, there are copy protection systems that will be very picky about things like this :-(

**Comment by hackgrid** [comment]
AsahiLina got many steam games running since at least March, so somehow it is possible already :-)

https://www.youtube.com/watch?v=CJSfFzsU75g

**Comment by Matthew-Bradley** [comment]
> AsahiLina got many steam games running since at least March, so somehow it is possible already :-)
> 
> https://www.youtube.com/watch?v=CJSfFzsU75g

As far as I know that was on one of the (broken) 4K page kernels.

**Comment by hackgrid** [comment]
Haha, seemed pretty functional to me :D What was broken?
But nevermind, I just mentioned it, because the consensus in this issue was, that this should be not possible, but it seems it is already! :)

**Comment by Matthew-Bradley** [comment]
> Haha, seemed pretty functional to me :D What was broken? But nevermind, I just mentioned it, because the consensus in this issue was, that this should be not possible, but it seems it is already! :)

The IOMMU. Apple's IOMMU hardware only works with 16K pages, and nobody (especially asahi) is willing to ship a kernel with a broken IOMMU.

People have known since the beginning that running a broken 4K kernel would let you run 4K windows emulation. Most of this thread exists because shipping this solution to users will not happen.

(The asahi folk have also decided not to upstream their attempt at running the 16K IOMMU with a 4K kernel. It's far to awful to ever be accepted into mainline, and they'd rather see 16K support added to applications.)

**Comment by AndreRH** [comment]
I thought they work on 4K support for the GPU according to https://asahilinux.org/2023/03/road-to-vulkan/
Anyway, with a 4K support Hangover should run as well as on a 64-Bit Raspberry Pi I assume.

**Comment by Matthew-Bradley** [comment]
> I thought they work on 4K support for the GPU according to https://asahilinux.org/2023/03/road-to-vulkan/ 

They worked on 4K support for *only* the GPU. The GPU now works on a (broken) 4K kernel, the IOMMU (still) does not.

**Comment by hackgrid** [comment]
Marcan seems very passionate about gaming on Asahi Linux, so I guess everything will work out, since many things already work or their team is working on that for a long time, so let's just see what the near future brings us :)

**Comment by hackgrid** [comment]
Just want to mention there is dev work on an interesting solution going on, via a microvm + 4K kernel:

https://bahn.social/@slp@fosstodon.org/111094390741959319

**Comment by hackgrid** [comment]
This is awesome now:

https://sinrega.org/2023-10-06-using-microvms-for-gaming-on-fedora-asahi/

Easy install of Fex/MicroVM stuff with native gpu performance for all! :)

**Comment by volkertb** [comment]
Have you read [Alyssa Rosenzweig's recent blog post](https://rosenzweig.io/blog/aaa-gaming-on-m1.html)?

The Asahi team got some AAA games to work in Asahi Linux on Apple Silicon Mac hardware, with actually playable performance.

The Asahi Linux developers used a MicroVM called [muvm](https://github.com/AsahiLinux/muvm) to get around the page size problem.

How much of all this stuff could be reused in the Hangover project? Are you guys already cooperating with the Asahi Linux developers?

**Comment by darkbasic** [comment]
Yeah it's an incredibly interesting solution for a problem that would otherwise be huge with the Linux kernel.

**Comment by AndreRH** [comment]
Might be as easy as unpacking the debian11 package somewhere and run `muvm /path/to/wine`
I just can't test as I don't have the hardware

**Comment by AndreRH** [comment]
There were some adjustments in Wine lately, please retest with Hangover 10.6 to see if it works now.

**Comment by hackgrid** [comment]
Why is Hangover even needed anymore? Since Wine 10.0 it can natively + run
FexEmu inside Wine for emulation if needed. What is the different from
Vanilla Wine to Hangover now?

André Zwing ***@***.***> schrieb am Mo., 21. Apr. 2025, 18:21:

**Comment by owlshrimp** [comment]
> Why is Hangover even needed anymore? Since Wine 10.0 it can natively + run
> FexEmu inside Wine for emulation if needed. What is the different from
> Vanilla Wine to Hangover now?
> 
> André Zwing ***@***.***> schrieb am Mo., 21. Apr. 2025, 18:21:
> […](#)

afaik this is essentially asking why we need hangover when we have hangover?

**Comment by darkbasic** [comment]
@hackgrid this is the recap from ChatGPT:

![Image](https://github.com/user-attachments/assets/eb39997a-4775-489e-bb3a-b593f6ad8f30)

**Comment by owlshrimp** [comment]
> [@hackgrid](https://github.com/hackgrid) this is the recap from ChatGPT:
> 
> ![Image](https://github.com/user-attachments/assets/eb39997a-4775-489e-bb3a-b593f6ad8f30)

That looks like it's almost totally wrong. TBH I wouldn't expect ChatGPT to get something like this right, ESPECIALLY since hangover is a component meant to work within wine. Most of that seems to be made up based on a nonexistant distinction between the two.

**Comment by hackgrid** [comment]
What are talking about, Hangover already uses Fexemu. And I still not get
what the difference to vanilla wine, sounds to me both have the same
features?

owlshrimp ***@***.***> schrieb am Di., 6. Mai 2025, 10:46:

**Comment by owlshrimp** [comment]
> What are talking about, Hangover already uses Fexemu. And I still not get
> what the difference to vanilla wine, sounds to me both have the same
> features?
> 
> owlshrimp ***@***.***> schrieb am Di., 6. Mai 2025, 10:46:
> […](#)

Wine and hangover are not something you can compare with each other. They're different parts of the **same software** so of course they have identical features. They're both parts of the same thing.

The thing that Hangover lets you do, as a component of wine, is run the bulk of the wine code outside the emulator. This means you don't have to pay the emulation overhead for that portion of the wine code. This is explained on the front page at https://github.com/AndreRH/hangover/

**Comment by hackgrid** [comment]
Yes, that is ARM64EC included in vanilla wine. So what is the difference to
Hangover now, you dont need it anymore?

owlshrimp ***@***.***> schrieb am Mi., 7. Mai 2025, 02:42:

**Comment by owlshrimp** [comment]
> Yes, that is ARM64EC included in vanilla wine. So what is the difference to
> Hangover now, you dont need it anymore?
> 
> owlshrimp ***@***.***> schrieb am Mi., 7. Mai 2025, 02:42:
> […](#)

If you don't use hangover with wine on arm, you're going to pay higher overhead and worse performance.

**Comment by stefand** [comment]
There is relatively little difference nowadays compared to what's in upstream Wine with ARM64EC + xtajit. I'm a bit out of the loop though, so the following might get some details wrong:

The original hangover built by me in 2017 was done before any windows x86-on-arm emulation was announced by Microsoft and used a somewhat different concept than how things work nowadays: I approached it as running x86 windows applications on arm Windows, with the fact that "arm windows" was provided by Wine just being an implementation detail. I wrote a pile of thunk DLLs (with a template generated by a script and details fixed up manually).

My aim at the time wasn't necessarily to build a long term solution but to figure out what the performance impact was going to be vs running all of Wine in qemu userspace emulation. The rough answer to that was "relatively little". If your CPU emulator is too slow to run application + wine, it is going to run too slow to run your application alone, even if you pull Wine out of the emulator. The two exceptions to that were 3D drivers and sound processing - pulling those out of the emulator made a big difference and is going to make a difference even with faster emulators. But few other DLLs or system libraries do a lot of CPU work themselves.

Upstream Wine didn't like my thunk approach very much, and the thunks were going to run into issues with the Windows API. E.g. I originally tried to thunk comctl32 and comdlg32, but the nature of WM_NOTIFY messages and wndproc hooking made it impossible to know what data is passed between comctl32 controls and the application.

Microsoft's own emulator for 32 bit x86 on arm64 combined with Anti-cheat software peeking into the x86 on x86_64 layers meant that in the long run Wine's wow64 had to become a lot more like Windows' own wow64 with thunking being done purely in userspace (compared to Linux/Mac, where the 32-to-64 transition happens at the userspace-kernel boundary). When Microsoft introduced the ARM64EC ABI for x86_64 on arm64 it gave Wine a good reason to replicate this system, which involved a lot of contributions to clang and mingw. If Microsoft had not built/defined ARM64EC, adding a system like this to upstream compilers would have been a tough sell. For a few years CrossOver ran x86 on x64 on MacOS with a custom compiler - see https://marc.info/?l=wine-devel&m=157610999725233&w=4 for technical details.

Once Wine's "new" wow64 became somewhat usable, André ditched my old hangover thunks and instead started providing an xtajit.dll that takes over the role of wow64cpu.dll to run 32 bit x86 on 64 bit arm. Like its early Windows counterpart, it did not run 64 bit x86 on arm64. This became possible when ARM64EC support started taking shape. For a while, Hangover provided an easier way to build and run the work in progress code, but as you've observed, it is in pretty good shape in upstream Wine and mingw nowadays - so you probably don't need this hangover repo and can build upstream clang, clang-mingw and Wine.

The other nice side effect of "new" wow64 is that it can run 32 bit Windows applications without needing 32 bit Linux libraries - so in the near-ish future you can probably tell your Linux multilib .so's goodbye, unless you need other 32 bit Linux software.

**Comment by hackgrid** [comment]
Wow, thanks Stefan Dösinger for your detailed explainations! :)

Stefan Dösinger ***@***.***> schrieb am Mi., 7. Mai 2025,
14:10:

**Comment by AndreRH** [comment]
First, this should have been a discussion or at least a separate issue instead of hijacking a 5 yo issue about something else...
Second, I wonder if I could delete the README, 'cause it seems nobody actually reads it.
A lot of things are upstream by now, but some things are not.
Hangover has:

- prebuilt packages
- box64cpu (though there's currently some effort to get that into Box64 [upstream](https://github.com/ptitSeb/box64/pull/2619))
- comfort defaults
- thread suspension support
- build fixes
- a good foundation to try out other emulators as dlls, like maybe the planned Box64 based ARM64EC module
- other interesting tricks

So either this doesn't get recognized or you don't read/think before you comment on the internet.

**Comment by hackgrid** [comment]
Thanks! The readme mentions nothing of these things though!
My original question was "ARM64EC is included in vanilla wine. So what is the difference to Hangover now, you dont need it anymore?" after all.

**Comment by rlippmann** [comment]
> Thanks! The readme mentions nothing of these things though! My original question was "ARM64EC is included in vanilla wine. So what is the difference to Hangover now, you dont need it anymore?" after all.

ARMEC is included in Wine, but:

- it is difficult to build (you can't just run configure).  you need a customized mingw toolchain
- the actual x86 emulation isn't implemented in vanilla wine (you need to use FEX to build an emulation DLL
- in terms of Apple Silicon, there is still the 16k vs 4k page size issues.  Windows ARM actually uses 64k page sizes.
- also, in terms of Apple Silicon, Windows expects certain things mapped at certain memory addresses, and Apple Silicon will not let user code map that low in memory

**Comment by hackgrid** [comment]
Thank you!
On Asahi Linux you can already use native ARM Wine with 64k page size and play 4K Windows games :)

**Comment by rlippmann** [comment]
> Thank you! On Asahi Linux you can already use native ARM Wine with 64k page size and play 4K Windows games :)

Can you?  I haven't tried.

You used to need to use muvm to do it (so, you were actually running FEX in an emulator).  But that was Steam for linux.

**Comment by hackgrid** [comment]
Yes, it runs native without muvm.

Check out
https://www.reddit.com/r/AsahiGaming/comments/1k1hmnf/red_dead_redemption_2_vulkan_on_winearm64ec/
https://www.reddit.com/r/AsahiLinux/comments/1k32fs5/wine_arm64ec_experimental_rpm_build_and_setup/

**Comment by stefand** [comment]
Wine's support for mostly working on larger pages sizes is relatively new - I think it was added somewhen in April. As far as I understand there will always be some pitfalls. If an application, e.g. a copy protection system, tries hard enough it can do things we can't fake.

Windows on ARM does *not* use 64k page sizes. It is 4k at the kernel level, but NtAllocateVirtualMemory only allows 64k aligned addresses. Smaller chunks are exposed to user space in the details though: PE files can have 4kb sections that are mapped accordingly, VirtualProtect allows you to change protections with 4k granularity, and if you mmap a, say, 25 kb file, you get a 64kb block of memory of which 28kb are accessible and the rest marked noaccess.

But yes, the fact that Microsoft makes it harder to use 4kb allocations makes things easier for Wine than running generic Linux software. I think the 64k restrictions are a result of early Windows NT supporting Alpha CPUs, which had 64k pages.

**Comment by AndreRH** [comment]
AFAIK this is fixed
